src/storage/compression/roaring.cpp

```diff
@@ -0,0 +1,1436 @@
+			do {
+				if (array_index >= count || scanned_count + to_scan < array[array_index]) {
+					// None of the bits we're scanning are 0, no action required
+					break;
+				}
+
+				// At least one of the entries to scan is set
+				for (; array_index < count; array_index++) {
+					if (array[array_index] >= scanned_count + to_scan) {
+						break;
+					}
+					if (array[array_index] < scanned_count) {
+						continue;
+					}
+					auto index = array[array_index] - scanned_count;
+					result_mask.SetInvalid(result_offset + index);
+				}
+			} while (false);
+			scanned_count += to_scan;
+		} else {
+			// We are mapping non-nulls
+			do {
+				if (array_index >= count || scanned_count + to_scan < array[array_index]) {
+					// None of the bits we're scanning are set, set everything to 0 directly
+					idx_t start = result_offset;
+					idx_t end = start + to_scan;
+					SetInvalidRange(result_mask, start, end);
+					break;
+				}
+
+				if (array_index + to_scan < count && array[array_index] == scanned_count &&
+				    array[array_index + to_scan] == scanned_count + to_scan) {
+					// All bits are set, no action required
+					break;
+				}
+
+				idx_t i = 0;
+				while (i < to_scan) {
+					// Determine the next valid position within the scan range, if available
+					idx_t valid_pos = (array_index < count) ? array[array_index] : scanned_count + to_scan;
+					idx_t valid_start = MinValue<idx_t>(valid_pos - scanned_count, to_scan);
+
+					if (i < valid_start) {
+						// These bits are all set to 0
+						idx_t start = result_offset + i;
+						idx_t end = start + (valid_start - i);
+						SetInvalidRange(result_mask, start, end);
+						i = valid_start;
+					}
+
+					if (i == valid_start && i < to_scan && array_index < count) {
+						// This bit is already set, no action required
+						i++;
+						array_index++;
+					}
+				}
+			} while (false);
+			scanned_count += to_scan;
+		}
+	}
+
+	void Skip(idx_t to_skip) override {
+		idx_t end = scanned_count + to_skip;
+		while (array_index < count && array[array_index] < end) {
+			array_index++;
+		}
+		// In case array_index has already reached count
+		scanned_count = end;
+	}
+
+	void Verify() const override {
+#ifdef DEBUG
+		idx_t index = 0;
+		for (idx_t i = 0; i < count; i++) {
+			D_ASSERT(!i || array[i] > index);
+			index = array[i];
+		}
+#endif
+	}
+
+public:
+	uint16_t *array;
+	const idx_t count;
+	idx_t array_index = 0;
+};
+
+struct BitsetContainerScanState : public ContainerScanState {
+public:
+	BitsetContainerScanState(idx_t container_index, idx_t count, validity_t *bitset)
+	    : ContainerScanState(container_index, count), bitset(bitset) {
+	}
+
+public:
+	void ScanPartial(Vector &result, idx_t result_offset, idx_t to_scan) override {
+		if ((to_scan % ValidityMask::BITS_PER_VALUE) == 0 && (scanned_count % ValidityMask::BITS_PER_VALUE) == 0) {
+			ValidityUncompressed::AlignedScan(reinterpret_cast<data_ptr_t>(bitset), scanned_count, result, to_scan);
```

Comment:
We need to use result_offset here, or check if result_offset == 0 before calling aligned scan